pkg/linode-bs/controllerserver.go

```diff
@@ -120,47 +102,48 @@ func (linodeCS *LinodeControllerServer) CreateVolume(ctx context.Context, req *c
 		volumeContext[LuksKeySizeAttribute] = req.Parameters[LuksKeySizeAttribute]
 	}
 
-	tags := req.Parameters[VolumeTags]
-
-	if len(volumes) != 0 {
-		if len(volumes) > 1 {
-			return nil, status.Error(codes.AlreadyExists, fmt.Sprintf("duplicate volume %q exists", volumeName))
-		}
-		volume := volumes[0]
-		if int64(volume.Size*gigabyte) != size {
-			return nil, status.Error(codes.AlreadyExists, fmt.Sprintf("invalid option requested size: %d", size))
-		}
+	targetSizeGB := int(size / gigabyte)
 
-		key := common.CreateLinodeVolumeKey(volume.ID, volume.Label)
-
-		klog.V(4).Info("volume already created")
-		return &csi.CreateVolumeResponse{
-			Volume: &csi.Volume{
-				VolumeId:      key.GetVolumeKey(),
-				CapacityBytes: int64(volume.Size * gigabyte),
-				VolumeContext: volumeContext,
-			},
-		}, nil
-	}
-
-	var vol *linodego.Volume
-	volumeSizeGB := int(size / gigabyte)
-
-	if sourceVolumeInfo != nil {
-		// Clone the volume
-		vol, err = linodeCS.cloneLinodeVolume(ctx, volumeName, volumeSizeGB, sourceVolumeInfo.VolumeID)
-	} else {
-		// Create the volume from scratch
-		vol, err = linodeCS.createLinodeVolume(ctx, volumeName, volumeSizeGB, tags)
+	// Attempt to get info about the source volume.
+	// sourceVolumeInfo will be null if no content source is defined.
+	contentSource := req.GetVolumeContentSource()
+	sourceVolumeInfo, err := linodeCS.attemptGetContentSourceVolume(
+		ctx,
+		contentSource,
+	)
+	if err != nil {
+		return nil, err
 	}
 
-	// Error handling for the above function calls
+	// Attempt to create the volume while respecting idempotency
+	vol, err := linodeCS.attemptCreateLinodeVolume(
+		ctx,
+		volumeName,
+		targetSizeGB,
+		req.Parameters[VolumeTags],
+		sourceVolumeInfo,
+	)
 	if err != nil {
 		return nil, err
 	}
 
-	if err != nil {
-		return nil, status.Error(codes.Internal, err.Error())
+	// Attempt to resize the volume if necessary
+	if vol.Size != targetSizeGB {
+		klog.V(4).Infoln("resizing volume", map[string]interface{}{
+			"volume_id": vol.ID,
+			"old_size":  vol.Size,
+			"new_size":  targetSizeGB,
+		})
+
+		if err := linodeCS.CloudProvider.ResizeVolume(ctx, vol.ID, targetSizeGB); err != nil {
+			return nil, status.Errorf(codes.Internal, "failed to resize cloned volume (%d): %s", targetSizeGB, err)
+		}
+	}
+
+	// Always wait for the volume to be active
+	if _, err := linodeCS.CloudProvider.WaitForVolumeStatus(
+		ctx, vol.ID, linodego.VolumeActive, waitTimeout); err != nil {
```

Comment:
Would it make sense to use a separate timeout const for clone operations? I'm hesitant because clone runtime can vary drastically depending on the contents of the disk (?).

Reply:
>Seems to make a lot of sense to use a different timeout for clone as I'd imagine (hope) a typical create volume operation would be much faster than cloning a multi GiB volume.

Sounds good, I'll implement that when I find some time.

>What happens in the case that the timeout gets hit - we return an error but the operation continues in the Linode API? Is there some kind of reconciliation elsewhere that'd account for the operation finishing?

So if the timeout is hit while [polling for volume readiness](https://github.com/linode/linode-blockstorage-csi-driver/pull/142/files#diff-a45fd181b265bc173ff27928709318c12e64ee8eedb87832081ef107dbc96197R144-R147), an error will be returned but the volume will **not** be destroyed. This error will implicitly trigger the volume creation request to be re-run, in which case [this bit of logic](https://github.com/linode/linode-blockstorage-csi-driver/pull/142/files#diff-a45fd181b265bc173ff27928709318c12e64ee8eedb87832081ef107dbc96197R560-R578) will discover the volume created during the previous run and will continue polling for readiness.